packages/client-proxy/internal/edge/sandboxes/catalog_redis.go

```diff
@@ -0,0 +1,121 @@
+package sandboxes
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-redsync/redsync/v4"
+	"github.com/jellydator/ttlcache/v3"
+	"github.com/redis/go-redis/v9"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	catalogClusterLockName = "sandboxes-catalog-cluster-lock"
+	catalogRedisTimeout    = time.Second * 5
+
+	// this is just how long we are keeping sandbox in local cache so we don't have to query redis every time
+	// we don't want to go too high because then sbx can be run on different orchestrator, and we will not be able to find it
+	catalogRedisLocalCacheTtl = time.Second * 10
+)
+
+type RedisSandboxCatalog struct {
+	clusterMutex *redsync.Mutex
+	redisClient  *redis.Client
+
+	cache  *ttlcache.Cache[string, *SandboxInfo]
+	ctx    context.Context
+	tracer trace.Tracer
+}
+
+func NewRedisSandboxesCatalog(ctx context.Context, tracer trace.Tracer, redisClient *redis.Client, redisSync *redsync.Redsync) SandboxesCatalog {
+	clusterLockMutex := redisSync.NewMutex(catalogClusterLockName)
+
+	cache := ttlcache.New(ttlcache.WithTTL[string, *SandboxInfo](catalogRedisLocalCacheTtl))
+	go cache.Start()
+
+	return &RedisSandboxCatalog{
+		redisClient:  redisClient,
+		clusterMutex: clusterLockMutex,
+		tracer:       tracer,
+		cache:        cache,
+		ctx:          ctx,
+	}
+}
+
+func (c *RedisSandboxCatalog) GetSandbox(sandboxId string) (*SandboxInfo, error) {
+	spanCtx, span := c.tracer.Start(c.ctx, "sandbox-catalog-get")
+	defer span.End()
+
+	sandboxInfo := c.cache.Get(sandboxId)
+	if sandboxInfo != nil {
+		return sandboxInfo.Value(), nil
+	}
+
+	ctx, ctxCancel := context.WithTimeout(spanCtx, catalogRedisTimeout)
+	defer ctxCancel()
+
+	data, err := c.redisClient.Get(ctx, c.getCatalogKey(sandboxId)).Bytes()
+	if err != nil {
+		return nil, ErrSandboxNotFound
+	}
+
+	var info *SandboxInfo
+	err = json.Unmarshal(data, &info)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal sandbox info: %w", err)
+	}
+
+	err = c.StoreSandbox(sandboxId, info)
```

Comment:
why do you call store here?

Reply:
when it is not in local cache, we will change redis, if its there we will store it in local one and return it.